### PR TITLE
Update containerinstance client to the 2021-03-01 api

### PR DESCRIPTION
**Pull request by joseph-porter**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [x] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [x] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

This generation includes a new API version 2021-03-01: https://github.com/Azure/azure-rest-api-specs/pull/12825
As well as an update to a previous API version: https://github.com/Azure/azure-rest-api-specs/pull/12129

### CODE REVIEW

**Comment by JonathanCrd** [comment]
Hi, could you please provide the approved swagger PR?

**Comment by joseph-porter** [comment]
@JonathanCrd 
This generation includes a new API version 2021-03-01: Azure/azure-rest-api-specs#12825
As well as an update to a previous API version: Azure/azure-rest-api-specs#12129

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs
Comment (JonathanCrd) [inline sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs:98]: This seems like a breaking change: introducing `timestamps` optional param not at the end.

Please rearrange the params in the swagger file so `bool? timestamps` is at the end.

**Review by JonathanCrd — CHANGES_REQUESTED** [review]

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs
Comment (joseph-porter) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs:98]: @JonathanCrd The timestamps param is already at the end of the param list in the Swagger for the Containers_ListLogs operation (https://github.com/Azure/azure-rest-api-specs/blob/master/specification/containerinstance/resource-manager/Microsoft.ContainerInstance/stable/2021-03-01/containerInstance.json)

Could this be an issue with the autogen? How could we fix it?

Comment (joseph-porter) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs:98]: @JonathanCrd Following up here. Since the parameter in the Swagger is already in the correct order, what can I do to correct this? Since this is generated by the autogen, I assume I cannot manually change the order.

Comment (JonathanCrd) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs:98]: Hi, I'm asking how to resolve this, since this has to do with the autogen. 
I'll let you know what to do, or I'll approve the PR if there's nothing else for you to do.

Comment (JonathanCrd) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs:98]: @joseph-porter, this issue is specific to .NET generation, and the way to fix it is by adding a partial class with an overload that implements the old API, generally this is done by making all the parameters up to the last common parameter required.

For example, in this case it would be: 
```C#
public async Task<AzureOperationResponse<Logs>> ListLogsWithHttpMessagesAsync(string resourceGroupName, string containerGroupName, string containerName, int tail, Dictionary<string, List<string>> customHeaders, CancellationToken cancellationToken = default(CancellationToken))
```

In that way, any calls that provided a value form customHeaders will use this added method, but any new method calls, and any method calls that do not use customer headers will automatically use the new method.

Thanks to @markcowl for providing this solution :)

Comment (joseph-porter) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs:130]: @JonathanCrd  I have added an overloaded method with the old API.

**Review by JonathanCrd — APPROVED** [review]
Approving changes, fix the CI errors to merge.

**Comment by joseph-porter** [comment]
@JonathanCrd Thanks for starting the pipeline, looks like it failed again:

D:\a\1\.nuget\packages\microsoft.internal.netsdkbuild.mgmt.tools\0.12.0-dev.20210610.1\targets\common.targets(119,5): error MSB4181: The "MSBuild" task returned false but did not log an error. [D:\a\1\s\mgmt.proj]

**Comment by joseph-porter** [comment]
@JonathanCrd I'm not seeing anything helpful in CI pipeline output to explain the failures. How can we move forward and unblock the CI issues?

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs
Comment (markcowl) [inline sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs:82]: It looks like you added this method directly onto the generated Operation class.  What you need to do to make this customization is the following:

(1) This method signature appears in IContainersOperations, ContainersOperations, and ContainersOperationsExtensions, and all 3 of these would need the new overload(s)

(2) We don't edit the generated code, as these methods would go away after the next round of generation.  Instead,  since the generated classes and interfaces are partial, we add a new class with just the new method signatures in a partial class with the same name in src/Customizations.  This will allow this customnization to continue through successive updates.

(3) The build is failing because for the async method to return Task<AzureOperationResponse<Logs>> you need to return an AzureOperationResponse<Logs>, and it will automatically be wrapped.  To do this, you would either mark this method as not async, or you would await the result of the call to the other method:
``` C#
return await ListLogsWithHttpMessagesAsync(resourceGroupName, containerGroupName, containerName, tail: tail, timestamps: false, customHeaders: customHeaders, cancellationToken: cancellationToken);
```

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/IContainersOperations.cs
Comment (markcowl) [inline sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/IContainersOperations.cs:65]: per comment above, also need to add a new overload in a custom partial interface.

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperationsExtensions.cs
Comment (markcowl) [inline sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperationsExtensions.cs:86]: Per previous comment, this will also require a customization

**Review by markcowl — CHANGES_REQUESTED** [review]

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs
Comment (markcowl) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs:82]: Since you aren't actually awaiting any calls, it is fine to remove the async keyword, without making the change above.

Comment (joseph-porter) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperations.cs]: @markcowl I moved the overload into customizations as a partial class. The examples really helped!

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/IContainersOperations.cs
Comment (joseph-porter) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/IContainersOperations.cs:65]: Added overload in a custom partial interface

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperationsExtensions.cs
Comment (joseph-porter) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Generated/ContainersOperationsExtensions.cs:86]: Added overload in a custom partial interface

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Customizations/ContainersOperations.cs
Comment (markcowl) [inline sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Customizations/ContainersOperations.cs:61]: We want to make sure that this method overload is differentiated from the existing method overload while preserving backward compatibility.  In this case, the calls that would be broken by the new API are calls using sequential parameters that include 'customHeaders', so the proper signature is:

``` c#
public async Task<AzureOperationResponse<Logs>> ListLogsWithHttpMessagesAsync(string resourceGroupName, string containerGroupName, string containerName, int? tail, Dictionary<string, List<string>> customHeaders, CancellationToken cancellationToken = default(CancellationToken))
```
this should apply to ContainersOperationsExtensions and IContainersOperations as well.

**Review by markcowl — CHANGES_REQUESTED** [review]
One requested change - the failures in net-mgmt-ci look to be places where you need to re-record the tests.

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Customizations/ContainersOperations.cs
Comment (joseph-porter) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Customizations/ContainersOperations.cs]: Changed to call the generated method, also re-recorded the test sessions.

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Customizations/ContainersOperations.cs
Comment (joseph-porter) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Customizations/ContainersOperations.cs]: @markcowl, I applied this overload in ContainerOperations and IContainerOperations , but I left ContainersOperationsExtensions the same since it does not use customHeaders.

**Comment by joseph-porter** [comment]
@markcowl @yungezz Any updates here? 

We're trying to update the SDK to address a customer issue, so hoping to make quick progress here.

**Comment by markcowl** [comment]
@joseph-porter The mgmt tests are now passing.  Need to figure out why the data plane tests, which this PR shouldn't impaxct are affected.  @weshaggard any ideas?

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Customizations/ContainersOperationsExtensions.cs
Comment (markcowl) [inline sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Customizations/ContainersOperationsExtensions.cs:40]: the same change to the operation signature for IContainersOperations and ContainersOperations should be duplicated here as well.

that is:
``` c#
public static async Task<Logs> ListLogsAsync(this IContainersOperations operations, string resourceGroupName, string containerGroupName, string containerName, int? tail, CancellationToken cancellationToken = default(CancellationToken))
```

**Review by markcowl — CHANGES_REQUESTED** [review]
Need to propagate the method signature change to the extension class as well.  Otherwise, thsi is ready to go.

File: sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Customizations/ContainersOperationsExtensions.cs
Comment (joseph-porter) [reply sdk/containerinstance/Microsoft.Azure.Management.ContainerInstance/src/Customizations/ContainersOperationsExtensions.cs]: @markcowl Ah, I see. Changed to match!

**Review by markcowl — APPROVED** [review]

**Pull request merged**
